src/odemis/acq/align/fastem.py

```diff
@@ -18,22 +20,282 @@
+
+        :param future: (ProgressiveFuture) Acquisition future object, which can be cancelled.
+                        (Exception or None): Exception raised during the calibration or None.
+        """
+        self._scanner = scanner
+        self._multibeam = multibeam
+        self._descanner = descanner
+        self._detector = detector
+        self._dataflow = detector.data
+        self._stage = stage
+        self._ccd = ccd
+        self._beamshift = beamshift
+        self._det_rotator = det_rotator
+        self._future = future
+
+        self.calibrations = calibrations
+
+        # List of calibrations to be executed. Used for progress update.
+        self._calibrations_remaining = set(calibrations)
+
+        # keep track if future was cancelled or not
+        self._cancelled = False
+
+        if model.MD_CALIB not in self._beamshift.getMetadata().keys():
+            self._beamshift.updateMetadata({model.MD_CALIB: self._scanner.beamShiftTransformationMatrix.value})
+
+    def run(self):
+        """
+        Runs a set of calibration procedures.
+        :returns:
+            (None) If calibrations successful return None.
+        :raise:
+            Exception: If a calibration failed.
+            CancelledError: If the calibration was cancelled.
+        """
+
+        # Get the estimated time for all requested calibrations.
+        total_calibration_time = self.estimate_calibration_time()
+
+        # No need to set the start time of the future: it's automatically done when setting its state to running.
+        self._future.set_progress(end=time.time() + total_calibration_time)  # provide end time to future
+        logging.info("Starting calibrations, with expected duration of %f s", total_calibration_time)
+
+        try:
+            logging.debug("Starting calibration.")
+
+            # loop over calibrations in list (order in list is important!)
+            for calib in self.calibrations:
+                # TODO return a sub-future when implemented for calibrations
+                self.run_calibrations(calib)
+
+                # def _pass_future_progress(sub_f, start, end):
+                #     f.set_progress(start, end)
+
+                # TODO Connect the progress of the sub-future to the main future
+                # sf.add_update_callback(_pass_future_progress)
+                # sf.result()
+
+                # remove from list of calibrations when finished
+                self._calibrations_remaining.discard(calib)
+
+                # In case the calibrations was cancelled by a client, before the future returned,
+                # raise cancellation error.
+                if self._cancelled:
+                    raise CancelledError()
+
+                # Update the time left for the calibrations remaining
+                self._future.set_progress(end=time.time() + self.estimate_calibration_time())
+
+        except CancelledError:  # raised in try statement
```

Comment:
This comment `# raised in try statement` confuses me, if you have an except, it's always because the error was raised in the try statement, right? This comment makes it sound like this is a special case, but I do not see why it would be a special case.

Reply:
True, in the acquisition it was more hidden as it was raised in acquire_roa, which itself was called in the try statement. I will remove it.